<a href="https://colab.research.google.com/github/Kristian-Abad/Netflix-Movies-Project/blob/main/Netflix_Movies.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing libraries

In [102]:
import numpy as np
import pandas as pd

# Importing the data

The data comes from [this Kaggle dataset.](https://www.kaggle.com/datasets/amjaads/netflix-movies?resource=download)

In [103]:
data = pd.read_csv('Netflix_movies.csv')

# Exploring our data

In [104]:
data.head()

Unnamed: 0           movie_name  Duration  year  \
0           1    #FriendButMarried       102  2018   
1           4              #Selfie       125  2014   
2           5           #Selfie 69       119  2016   
3           7         #realityhigh        99  2017   
4           9  10 Days in Sun City        87  2017   

                                             genre          director  \
0    Dramas, International Movies, Romantic Movies     Rako Prijanto   
1           Comedies, Dramas, International Movies    Cristina Jacob   
2           Comedies, Dramas, International Movies    Cristina Jacob   
3                                         Comedies  Fernando Lebrija   
4  Comedies, International Movies, Romantic Movies         Adze Ugah   

                                              actors        country  rating  \
0  Adipati Dolken, Vanesha Prescilla, Rendi Jhon,...      Indonesia     7.0   
1  Flavia Hojda, Crina Semciuc, Olimpia Melinte, ...        Romania     6.1   
2  Maia Morgenstern, Olimpia Melinte, Crina Semci...        Romania     6.3   
3  Nesta Cooper, Kate Walsh, John Michael Higgins...  United States     5.2   
4  Ayo Makun, Adesua Etomi, Richard Mofe-Damijo, ...   South Africa     5.3   

    enter_in_netflix  
0       May 21, 2020  
1       June 1, 2019  
2       June 1, 2019  
3  September 8, 2017  
4   October 18, 2019

In [105]:
data.loc[:,'Unnamed: 0']

0          1
1          4
2          5
3          7
4          9
        ... 
3318    3892
3319    3893
3320    3894
3321    3895
3322    3896
Name: Unnamed: 0, Length: 3323, dtype: int64

We're going to need to deal with this column later.

In [106]:
data.shape

(3323, 10)

Okay so we're dealing with a dataframe with 3,323 records across 10 variables.

In [107]:
data.isna().any()

Unnamed: 0          False
movie_name          False
Duration            False
year                False
genre               False
director            False
actors              False
country             False
rating              False
enter_in_netflix    False
dtype: bool

Looks like we are fortunate enough to not have any missing values.

In [108]:
data = data.rename(columns={"Unnamed: 0" : "index"})
data.head()

index           movie_name  Duration  year  \
0      1    #FriendButMarried       102  2018   
1      4              #Selfie       125  2014   
2      5           #Selfie 69       119  2016   
3      7         #realityhigh        99  2017   
4      9  10 Days in Sun City        87  2017   

                                             genre          director  \
0    Dramas, International Movies, Romantic Movies     Rako Prijanto   
1           Comedies, Dramas, International Movies    Cristina Jacob   
2           Comedies, Dramas, International Movies    Cristina Jacob   
3                                         Comedies  Fernando Lebrija   
4  Comedies, International Movies, Romantic Movies         Adze Ugah   

                                              actors        country  rating  \
0  Adipati Dolken, Vanesha Prescilla, Rendi Jhon,...      Indonesia     7.0   
1  Flavia Hojda, Crina Semciuc, Olimpia Melinte, ...        Romania     6.1   
2  Maia Morgenstern, Olimpia Melinte, Crina Semci...        Romania     6.3   
3  Nesta Cooper, Kate Walsh, John Michael Higgins...  United States     5.2   
4  Ayo Makun, Adesua Etomi, Richard Mofe-Damijo, ...   South Africa     5.3   

    enter_in_netflix  
0       May 21, 2020  
1       June 1, 2019  
2       June 1, 2019  
3  September 8, 2017  
4   October 18, 2019

Since the now index column is non-monotonic let's just drop it.

In [109]:
data = data.drop(columns='index')
# Verifying we have dropped the 'index' column
data.head()

movie_name  Duration  year  \
0    #FriendButMarried       102  2018   
1              #Selfie       125  2014   
2           #Selfie 69       119  2016   
3         #realityhigh        99  2017   
4  10 Days in Sun City        87  2017   

                                             genre          director  \
0    Dramas, International Movies, Romantic Movies     Rako Prijanto   
1           Comedies, Dramas, International Movies    Cristina Jacob   
2           Comedies, Dramas, International Movies    Cristina Jacob   
3                                         Comedies  Fernando Lebrija   
4  Comedies, International Movies, Romantic Movies         Adze Ugah   

                                              actors        country  rating  \
0  Adipati Dolken, Vanesha Prescilla, Rendi Jhon,...      Indonesia     7.0   
1  Flavia Hojda, Crina Semciuc, Olimpia Melinte, ...        Romania     6.1   
2  Maia Morgenstern, Olimpia Melinte, Crina Semci...        Romania     6.3   
3  Nesta Cooper, Kate Walsh, John Michael Higgins...  United States     5.2   
4  Ayo Makun, Adesua Etomi, Richard Mofe-Damijo, ...   South Africa     5.3   

    enter_in_netflix  
0       May 21, 2020  
1       June 1, 2019  
2       June 1, 2019  
3  September 8, 2017  
4   October 18, 2019

In [110]:
data[data.genre == 'Comedies'].head()

movie_name  Duration  year     genre  \
3                   #realityhigh        99  2017  Comedies   
61          A Bad Moms Christmas       104  2017  Comedies   
77   A Futile and Stupid Gesture       102  2018  Comedies   
107             A Thousand Words        92  2012  Comedies   
149  Adventures in Public School        86  2018  Comedies   

                   director  \
3          Fernando Lebrija   
61   Jon Lucas, Scott Moore   
77               David Wain   
107           Brian Robbins   
149            Kyle Rideout   

                                                actors                country  \
3    Nesta Cooper, Kate Walsh, John Michael Higgins...          United States   
61   Mila Kunis, Kristen Bell, Kathryn Hahn, Susan ...   United States, China   
77   Will Forte, Domhnall Gleeson, Martin Mull, Joe...          United States   
107  Eddie Murphy, Kerry Washington, Cliff Curtis, ...          United States   
149  Daniel Doheny, Judy Greer, Siobhan Williams, R...  Canada, United States   

     rating   enter_in_netflix  
3       5.2  September 8, 2017  
61      5.6  February 14, 2020  
77      6.8   January 26, 2018  
107     5.9       July 1, 2020  
149     5.8    August 15, 2018

I want to compare the count of movies that are comedies (or contain comedy elements), horror (or contain horror elements), and possibly both as indicated by the 'genre' column by year entered in Netflix. Here's how I would initially like to tackle this problem:


*   Create a subset of the data containing only comedies.
*   Create a subset of the data containing only horror.
*   Create a subset of the data containing both.
*   Combine the three subsets
*   Group by year entered in netflix and genre. Then count, sum as a new column called count
*   Export for visualization in Tableau







Let's find all the movies that at least 'Comedies' in the genre column. The [reference](https://thecodingbot.com/check-if-a-column-contains-specific-string-in-a-pandas-dataframe/) for the function used.

In [111]:
# Create a boolean filter of the records that contain 'Comedies' in the 'genre column'
comedies_filter = data.genre.str.contains('Comedies')

# Create a temporary subset of the data using the filter above
temp_subset = data[comedies_filter]

# Use the temporary subset to now filter out the "Horror" movies
data_comedies = temp_subset[~temp_subset.genre.str.contains('Horror')]

In [112]:
# Verifying that we should get no entries that have 'Horror' in the genre
data_comedies[data_comedies.genre.str.contains('Horror')]

Empty DataFrame
Columns: [movie_name, Duration, year, genre, director, actors, country, rating, enter_in_netflix]
Index: []

Likewise we can do the same operations for our other subsets of the data:

In [113]:
# Create a boolean filter of the records that contain 'Horror' in the 'genre column'
horror_filter = data.genre.str.contains('Horror')

# Create a temporary subset of the data using the filter above
temp_subset_2 = data[horror_filter]

# Use the temporary subset to now filter out the "Horror" movies
data_horror = temp_subset_2[~temp_subset_2.genre.str.contains('Comedies')]

In [114]:
data_horror.head()

movie_name  Duration  year  \
9          122        95  2019   
10  13 Cameras        90  2015   
11     13 Sins        93  2014   
13  14 Cameras        89  2018   
17        1920       143  2008   

                                             genre  \
9              Horror Movies, International Movies   
10    Horror Movies, Independent Movies, Thrillers   
11                        Horror Movies, Thrillers   
13                        Horror Movies, Thrillers   
17  Horror Movies, International Movies, Thrillers   

                      director  \
9              Yasir Al Yasiri   
10              Victor Zarcoff   
11                Daniel Stamm   
13  Scott Hussion, Seth Fuller   
17                Vikram Bhatt   

                                               actors        country  rating  \
9   Amina Khalil, Ahmed Dawood, Tarek Lotfy, Ahmed...          Egypt     7.1   
10  PJ McCabe, Brianne Moncrief, Sarah Baldwin, Ji...  United States     5.1   
11  Mark Webber, Rutina Wesley, Devon Graye, Tom B...  United States     6.3   
13  Neville Archambault, Amber Midthunder, Brytnee...  United States     4.5   
17  Rajneesh Duggal, Adah Sharma, Indraneil Sengup...          India     6.4   

     enter_in_netflix  
9        June 1, 2020  
10    August 13, 2016  
11   January 13, 2019  
13   October 27, 2018  
17  December 15, 2017

In [115]:
# Verifying that we should get no entries that have 'Horror' in the genre
data_horror[data_horror.genre.str.contains('Comedies')]

Empty DataFrame
Columns: [movie_name, Duration, year, genre, director, actors, country, rating, enter_in_netflix]
Index: []

In [116]:
# Create a boolean filter of the records that contain both 'Comedies' and 'Horror' in the 'genre column'
ch_filter = data.genre.str.contains('Horror') & data.genre.str.contains('Comedies')

# Create a temporary subset of the data using the filter above
data_ch = data[ch_filter]
data_ch.head()

movie_name  Duration  year  \
81     A Haunted House        86  2013   
1047  Girls With Balls        77  2019   
1112   Hantu Kak Limah       110  2018   
1135     He Never Died        97  2015   
1173          Holidays       106  2016   

                                              genre  \
81                          Comedies, Horror Movies   
1047  Comedies, Horror Movies, International Movies   
1112  Comedies, Horror Movies, International Movies   
1135                        Comedies, Horror Movies   
1173                        Comedies, Horror Movies   

                                               director  \
81                                       Michael Tiddes   
1047                                     Olivier Afonso   
1112                                       Mamat Khalid   
1135                                     Jason Krawczyk   
1173  Anthony Scott Burns, Nicholas McCarthy, Adam E...   

                                                 actors          country  \
81    Marlon Wayans, Essence Atkins, Cedric the Ente...    United States   
1047  Victor Artus Solaro, Manon Azem, Louise Blachè...  France, Belgium   
1112  Awie, Arab Khalid, Sharwani Ns, Tj Isa, Delima...         Malaysia   
1135  Henry Rollins, Booboo Stewart, Kate Greenhouse...           Canada   
1173  Madeleine Coghlan, Savannah Kennick, Ruth Brad...    United States   

      rating   enter_in_netflix  
81       5.0  February 21, 2020  
1047     3.8      July 26, 2019  
1112     6.3   January 31, 2019  
1135     6.4     March 18, 2016  
1173     5.1      July 15, 2016

Taking a quick peak into the shape of our subsets:

In [117]:
print('Comedies: ', data_comedies.shape,
      '\nHorror: ', data_horror.shape,'\n'
      'Comedies and Horror: ',data_ch.shape)

Comedies:  (987, 9) 
Horror:  (196, 9) 
Comedies and Horror:  (23, 9)


Let's change up the genres for each of the subsets.

In [118]:
data_comedies.loc[:,'genre'] = 'Comedies'
data_horror.loc[:,'genre'] = 'Horror'
data_ch.loc[:,'genre'] = 'Both'

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


Alright now let's combine all the subsets.

In [119]:
# Combine all the subsets into a new dataframe
data_complete_ch = pd.concat([data_comedies,data_horror,data_ch])
data_complete_ch.head()

movie_name  Duration  year     genre           director  \
1              #Selfie       125  2014  Comedies     Cristina Jacob   
2           #Selfie 69       119  2016  Comedies     Cristina Jacob   
3         #realityhigh        99  2017  Comedies   Fernando Lebrija   
4  10 Days in Sun City        87  2017  Comedies          Adze Ugah   
5       10 jours en or        97  2012  Comedies  Nicolas Brossette   

                                              actors        country  rating  \
1  Flavia Hojda, Crina Semciuc, Olimpia Melinte, ...        Romania     6.1   
2  Maia Morgenstern, Olimpia Melinte, Crina Semci...        Romania     6.3   
3  Nesta Cooper, Kate Walsh, John Michael Higgins...  United States     5.2   
4  Ayo Makun, Adesua Etomi, Richard Mofe-Damijo, ...   South Africa     5.3   
5  Franck Dubosc, Claude Rich, Marie Kremer, Math...         France     5.8   

    enter_in_netflix  
1       June 1, 2019  
2       June 1, 2019  
3  September 8, 2017  
4   October 18, 2019  
5       July 1, 2017

Let's do some finishing touches on the data by only getting the year that movies were added to Netflix and only

In [120]:
data_complete_ch.loc[:,'enter_in_netflix'] = pd.DatetimeIndex(data_complete_ch.enter_in_netflix).year
data_complete_ch.head()

movie_name  Duration  year     genre           director  \
1              #Selfie       125  2014  Comedies     Cristina Jacob   
2           #Selfie 69       119  2016  Comedies     Cristina Jacob   
3         #realityhigh        99  2017  Comedies   Fernando Lebrija   
4  10 Days in Sun City        87  2017  Comedies          Adze Ugah   
5       10 jours en or        97  2012  Comedies  Nicolas Brossette   

                                              actors        country  rating  \
1  Flavia Hojda, Crina Semciuc, Olimpia Melinte, ...        Romania     6.1   
2  Maia Morgenstern, Olimpia Melinte, Crina Semci...        Romania     6.3   
3  Nesta Cooper, Kate Walsh, John Michael Higgins...  United States     5.2   
4  Ayo Makun, Adesua Etomi, Richard Mofe-Damijo, ...   South Africa     5.3   
5  Franck Dubosc, Claude Rich, Marie Kremer, Math...         France     5.8   

   enter_in_netflix  
1              2019  
2              2019  
3              2017  
4              2019  
5              2017

In [121]:
data_complete_ch = data_complete_ch.drop(columns = ['director','actors']).rename(columns={'enter_in_netflix':'netflix_year'})
data_complete_ch.head()

movie_name  Duration  year     genre        country  rating  \
1              #Selfie       125  2014  Comedies        Romania     6.1   
2           #Selfie 69       119  2016  Comedies        Romania     6.3   
3         #realityhigh        99  2017  Comedies  United States     5.2   
4  10 Days in Sun City        87  2017  Comedies   South Africa     5.3   
5       10 jours en or        97  2012  Comedies         France     5.8   

   netflix_year  
1          2019  
2          2019  
3          2017  
4          2019  
5          2017


To do:
*   Aggregate data by genre, netflix year
*   Make a new column called count that gives the count of movies for each genre for each year
*   Make a new column called avg rating that gives the average rating for each genre for each year
*   Export to csv



In [122]:
data_complete_ch_mod = data_complete_ch.copy()
data_complete_ch_mod['count'] = data_complete_ch_mod.groupby(['genre','netflix_year'])['genre'].transform('count')
data_complete_ch_mod

movie_name  Duration  year     genre  \
1                           #Selfie       125  2014  Comedies   
2                        #Selfie 69       119  2016  Comedies   
3                      #realityhigh        99  2017  Comedies   
4               10 Days in Sun City        87  2017  Comedies   
5                    10 jours en or        97  2012  Comedies   
...                             ...       ...   ...       ...   
2628                 The Babysitter        85  2017      Both   
2633                        The Bar       102  2017      Both   
2844                    The Mansion       100  2017      Both   
3081  Tremors 6: A Cold Day in Hell        98  2018      Both   
3098       Tucker and Dale vs. Evil        89  2010      Both   

                                           country  rating  netflix_year  \
1                                          Romania     6.1          2019   
2                                          Romania     6.3          2019   
3                                    United States     5.2          2017   
4                                     South Africa     5.3          2019   
5                                           France     5.8          2017   
...                                            ...     ...           ...   
2628                                 United States     6.3          2017   
2633                              Spain, Argentina     6.3          2017   
2844                               France, Belgium     4.7          2018   
3081                   United States, South Africa     5.0          2018   
3098  Canada, United States, India, United Kingdom     7.5          2019   

      count  
1       300  
2       300  
3       155  
4       300  
5       155  
...     ...  
2628      4  
2633      4  
2844     10  
3081     10  
3098      4  

[1206 rows x 8 columns]

In [137]:
data_complete_ch_mod_2 = data_complete_ch_mod.copy()
data_complete_ch_mod_2['avg_rating'] = data_complete_ch_mod_2.groupby(['genre','netflix_year'])['rating'].transform('mean')

movie_name  Duration  year     genre  \
2489                     Splatter        29  2009    Horror   
1671  Mad Ron's Prevues from Hell        84  1987    Horror   
554              Casa de mi Padre        84  2012  Comedies   
2412            Shrek the Musical       130  2013  Comedies   
1030                    Get Santa       103  2014  Comedies   
...                           ...       ...   ...       ...   
1491              Killing Gunther        93  2017  Comedies   
3061          Tootsies & The Fake       108  2019  Comedies   
3060                      Tootsie       116  1982  Comedies   
1522           Kuch Kuch Hota Hai       185  1998  Comedies   
3240           Where the Money Is        89  2000  Comedies   

                                             country  rating  netflix_year  \
2489                                   United States     3.1          2009   
1671                                   United States     6.1          2010   
554                            United States, Mexico     5.5          2012   
2412                                   United States     7.1          2013   
1030                   United Kingdom, United States     6.2          2014   
...                                              ...     ...           ...   
1491                                   United States     4.8          2020   
3061                                        Thailand     5.9          2020   
3060                                   United States     7.4          2020   
1522                                           India     7.6          2020   
3240  Germany, United States, United Kingdom, Canada     6.2          2020   

      count  avg_rating  
2489      1    3.100000  
1671      1    6.100000  
554       1    5.500000  
2412      1    7.100000  
1030      1    6.200000  
...     ...         ...  
1491    247    6.019433  
3061    247    6.019433  
3060    247    6.019433  
1522    247    6.019433  
3240    247    6.019433  

[1206 rows x 9 columns]

Here's the code to verify both aggregations:

In [140]:
data_complete_ch.groupby(['genre','netflix_year']).count()

movie_name  Duration  year  country  rating
genre    netflix_year                                             
Both     2016                   2         2     2        2       2
         2017                   4         4     4        4       4
         2018                  10        10    10       10      10
         2019                   4         4     4        4       4
         2020                   3         3     3        3       3
Comedies 2012                   1         1     1        1       1
         2013                   1         1     1        1       1
         2014                   1         1     1        1       1
         2015                  10        10    10       10      10
         2016                  39        39    39       39      39
         2017                 155       155   155      155     155
         2018                 233       233   233      233     233
         2019                 300       300   300      300     300
         2020                 247       247   247      247     247
Horror   2009                   1         1     1        1       1
         2010                   1         1     1        1       1
         2015                   2         2     2        2       2
         2016                   5         5     5        5       5
         2017                  18        18    18       18      18
         2018                  51        51    51       51      51
         2019                  80        80    80       80      80
         2020                  38        38    38       38      38

In [139]:
data_complete_ch.groupby(['genre','netflix_year'])['rating'].mean()

genre     netflix_year
Both      2016            5.750000
          2017            5.700000
          2018            5.220000
          2019            6.300000
          2020            5.500000
Comedies  2012            5.500000
          2013            7.100000
          2014            6.200000
          2015            5.560000
          2016            5.669231
          2017            5.902581
          2018            6.047639
          2019            5.864667
          2020            6.019433
Horror    2009            3.100000
          2010            6.100000
          2015            6.100000
          2016            5.120000
          2017            5.172222
          2018            5.274510
          2019            5.177500
          2020            5.789474
Name: rating, dtype: float64

Now that we have our subset, let's export it

In [138]:
data_complete_ch_mod_2.to_csv('/content/comedies_horror_subset.csv')

Alright now our data is ready for Tableau Data Visualization!